lib/basedriver/image-helpers.js

```diff
@@ -0,0 +1,110 @@
+import { Buffer } from 'buffer';
+import { PNG } from 'pngjs';
+import B from 'bluebird';
+
+/**
+ * @typedef {Object} Region
+ * @property {number} left - The offset from the left side
+ * @property {number} top - The offset from the top
+ * @property {number} width - The width
+ * @property {number} height - The height
+ */
+
+const BYTES_IN_PIXEL_BLOCK = 4;
+
+/**
+ * Crop the image by given rectangle (use base64 string as input and output)
+ *
+ * @param {string} base64Image The string with base64 encoded image
+ * @param {Region} rect The selected region of image
+ * @return {string} base64 encoded string of cropped image
+ */
+async function cropBase64Image (base64Image, rect) {
+  const image = await base64ToImage(base64Image);
+  const croppedImage = cropImage(image, rect);
+  return await imageToBase64(croppedImage);
+}
+
+/**
+ * Create a pngjs image from given base64 image
+ *
+ * @param {string} base64Image The string with base64 encoded image
+ * @return {PNG} The image object
+ */
+async function base64ToImage (base64Image) {
+  const imageBuffer = new Buffer(base64Image, 'base64');
+  return new B((resolve, reject) => {
+    const image = new PNG({filterType: 4});
+    image.parse(imageBuffer, (err, image) => { // eslint-disable-line promise/prefer-await-to-callbacks
+      if (err) { return reject(err); }
+      resolve(image);
+    });
+  });
+}
+
+/**
+ * Create a base64 string for given image object
+ *
+ * @param {PNG} image The image object
+ * @return {string} The string with base64 encoded image
+ */
+async function imageToBase64 (image) {
+  return new B((resolve, reject) => {
+    const chunks = [];
+    image.pack().on('data', (chunk) => chunks.push(chunk)).on('end', () => {
+      resolve(Buffer.concat(chunks).toString('base64'));
+    }).on('error', (err) => { // eslint-disable-line promise/prefer-await-to-callbacks
+      reject(err);
+    });
+  });
+}
+
+/**
+ * Crop the image by given rectangle
+ *
+ * @param {PNG} image The image to crop
+ * @param {Region} rect The selected region of image
+ * @return {PNG} Cropped image
+ */
+function cropImage (image, rect) {
+  const imageRect = {width: image.width, height: image.height};
+  const interRect = getRectIntersection(rect, imageRect);
+  if (isEmptyRect(interRect)) {
+    throw new Error(`Cannot crop ${JSON.stringify(rect)} from ${JSON.stringify(imageRect)} because the intersection between them is an empty rectangle`);
+  }
+
+  const firstVerticalPixel = interRect.top;
+  const lastVerticalPixel = interRect.top + interRect.height;
+
+  const firstHorizontalPixel = interRect.left;
+  const lastHorizontalPixel = interRect.left + interRect.width;
+
+  const croppedArray = [];
+  for (let y = firstVerticalPixel; y < lastVerticalPixel; y++) {
+    for (let x = firstHorizontalPixel; x < lastHorizontalPixel; x++) {
+      const firstByteIdxInPixelBlock = (imageRect.width * y + x) << 2;
```

Comment:
is there any reason to use the bitshift operator instead of simply multiplying by `BYTES_IN_PIXEL_BLOCK` which appears to do the same thing? maybe i don't understand image math :-)

Reply:
Yes, it should work in same way. But, I think bitshift should be faster, maybe I'm wrong, it's javascript :)